src/libsystemd/sd-json/json-util.c

```diff
@@ -144,3 +146,133 @@ int json_dispatch_path(const char *name, sd_json_variant *variant, sd_json_dispa
+
+        if (data.pid > PID_T_MAX || !pid_is_valid(data.pid))
+                return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is not a valid PID.", strna(name));
+
+        int local_boot_id = -1; /* tristate */
+        if (!sd_id128_is_null(data.boot_id)) {
+                sd_id128_t my_boot_id;
+
+                r = sd_id128_get_boot(&my_boot_id);
+                if (r < 0) {
+                        json_log(variant, flags | (FLAGS_SET(flags, SD_JSON_STRICT) ? 0 : SD_JSON_DEBUG), r, "Unable to get local boot ID to validate JSON field '%s': %m", strna(name));
+                        if (FLAGS_SET(flags, SD_JSON_STRICT))
+                                return r;
+                } else {
+                        local_boot_id = sd_id128_equal(data.boot_id, my_boot_id);
+                        if (!local_boot_id) {
+                                json_log(variant, flags | (FLAGS_SET(flags, SD_JSON_STRICT) ? 0 : SD_JSON_DEBUG), 0, "JSON field '%s' refers to non-local PID.", strna(name));
+                                if (FLAGS_SET(flags, SD_JSON_STRICT))
+                                        return -ESRCH;
+                        }
+                }
+        }
+
+        _cleanup_(pidref_done) PidRef np = PIDREF_NULL;
+        if (local_boot_id != 0) {
+                /* Try to acquire a pidfd – unless this is definitely not a local PID */
+                r = pidref_set_pid(&np, data.pid);
+                if (r < 0) {
+                        json_log(variant, flags | (FLAGS_SET(flags, SD_JSON_STRICT) ? 0 : SD_JSON_DEBUG), r, "Unable to get fd for PID in JSON field '%s': %m", strna(name));
+                        if (FLAGS_SET(flags, SD_JSON_STRICT))
+                                return r;
+                }
+        }
+
+        /* If the the PID is dead or we otherwise can't get a pidfd of it, then store at least the PID number */
+        if (!pidref_is_set(&np))
+                np = PIDREF_MAKE_FROM_PID(data.pid);
+
+        /* If the pidfd inode nr is specified, validate it or at least state */
+        if (data.fd_id > 0) {
+                if (np.fd >= 0) {
+                        r = pidref_acquire_pidfd_id(&np);
+                        if (r < 0 && !ERRNO_IS_NOT_SUPPORTED(r))
+                                return json_log(variant, flags, r, "Unable to get pidfd ID to validate JSON field '%s': %m", strna(name));
+
+                        if (data.fd_id != np.fd_id) {
+                                json_log(variant, flags | (FLAGS_SET(flags, SD_JSON_STRICT) ? 0 : SD_JSON_DEBUG), 0, "JSON field '%s' references PID with non-matching inode number.", strna(name));
+                                if (FLAGS_SET(flags, SD_JSON_STRICT))
+                                        return -ESRCH;
+                        }
+                } else if (local_boot_id != 0)
```

Comment:
I would probably debug log here that we didn't validate the pidfd id in this case